pydantic_ai_slim/pydantic_ai/agent/__init__.py

```diff
@@ -1497,35 +1170,42 @@ def _get_deps(self: Agent[T, OutputDataT], deps: T) -> T:
             return deps
 
     def _get_toolset(
-        self,
-        output_toolset: AbstractToolset[AgentDepsT] | None | _utils.Unset = _utils.UNSET,
-        additional_toolsets: Sequence[AbstractToolset[AgentDepsT]] | None = None,
+        self, additional: Sequence[AbstractToolset[AgentDepsT]] | None = None
     ) -> AbstractToolset[AgentDepsT]:
-        """Get the complete toolset.
+        """Get the combined toolset containing function tools registered directly to the agent and user-provided toolsets including MCP servers.
 
         Args:
-            output_toolset: The output toolset to use instead of the one built at agent construction time.
-            additional_toolsets: Additional toolsets to add.
+            additional: Additional toolsets to add.
         """
+        if some_tools := self._override_tools.get():
+            function_toolset = _AgentFunctionToolset(some_tools.value, max_retries=self._max_tool_retries)
+        else:
+            function_toolset = self._function_toolset
+
         if some_user_toolsets := self._override_toolsets.get():
             user_toolsets = some_user_toolsets.value
-        elif additional_toolsets is not None:
-            user_toolsets = [*self._user_toolsets, *additional_toolsets]
+        elif additional is not None:
+            user_toolsets = [*self._user_toolsets, *additional]
         else:
             user_toolsets = self._user_toolsets
 
-        all_toolsets = [self._function_toolset, *user_toolsets]
+        if user_toolsets:
+            toolset = CombinedToolset([function_toolset, *user_toolsets])
+        else:
+            toolset = function_toolset
 
         if self._prepare_tools:
-            all_toolsets = [PreparedToolset(CombinedToolset(all_toolsets), self._prepare_tools)]
+            toolset = PreparedToolset(toolset, self._prepare_tools)
 
-        output_toolset = output_toolset if _utils.is_set(output_toolset) else self._output_toolset
-        if output_toolset is not None:
-            if self._prepare_output_tools:
-                output_toolset = PreparedToolset(output_toolset, self._prepare_output_tools)
-            all_toolsets = [output_toolset, *all_toolsets]
+        return toolset
 
-        return CombinedToolset(all_toolsets)
+    @property
+    def toolset(self) -> AbstractToolset[AgentDepsT]:
```

Comment:
This will need to renamed (to `toolsets` returning a list?) once https://github.com/pydantic/pydantic-ai/pull/2366 is merged, as it adds an `@agent.toolset` decorator